### ISSUE TITLE
JWKS: load public keys from a well known endpoint

### ISSUE THREAD

**Issue by purpleKarrot** [PostgREST/postgrest#1130]
Please see https://auth0.com/docs/jwks

> Auth0 exposes a JWKS endpoint for each tenant, which is found at `https://YOUR_AUTH0_DOMAIN/.well-known/jwks.json`. This endpoint will contain the JWK used to sign all Auth0 issued JWTs for this tenant.

It would be great of PostgREST could load the public key from such an endpoint.

**Comment by ruslantalpa** [comment]
Or ... you install postgrest as a system service (https://docs.subzero.cloud/production-infrastructure/ubuntu-server/) and use systemd `ExecStartPre` directive to run one `curl` call that places that key in the file postgrest is configured to read...

**Comment by photz** [comment]
@purpleKarrot Did you get PostgREST to work with the JWK from Auth0 in general? I copied the first key in mytenant.auth0.com/.well-known/jwks.json into a file and set `jwt-secret = "@thefile"` in the config file as describe in the docs  [here](https://postgrest.org/en/v5.0/auth.html#asymmetric-keys).
Unfortunately PostgREST still always responds with
```
{
    "message": "JWSError JWSInvalidSignature"
}
```

**Comment by jcozain** [comment]
i was just about to ask something similar. does the jwk accepts the x5u parameter. i tried authenticating with firebase ID token (jwt signed with asymetrical RSA) and only could make it work with a node js library to transform the cert into a "standard" jwk [[jwk](https://www.npmjs.com/package/json-web-key)]. wich is not very maintainable because google has 2 public keys wich are going to change periodically.
 here are the keys (x5u compliant )
[google public key](https://www.googleapis.com/robot/v1/metadata/x509/securetoken@system.gserviceaccount.com)

**Comment by jcozain** [comment]
i made a nodejs script to fectch the certs and mint a jwks, the individual jwk work but when i put them both in a jwks it stops validating. im guessing this is not suported. i can get by because luckyly all of the jwt provided by google firebase had been of the same "kid" so im using just one key. any info would be welcome.

**Comment by photz** [comment]
@jcozain Yes, PostgREST seems to only support single JSON Web Keys at the moment, not JSON Web Key Sets.
I am using `jq` to extract the first JWK and remove the properties that hs-jose currently seems to be incompatible with (namely x5c and x5t).
```
wget --quiet -O - https://yourcompany.eu.auth0.com/.well-known/jwks.json \
    | jq " {
               alg: .keys[0].alg,
               kty: .keys[0].kty,
               use: .keys[0].use,
               n: .keys[0].n,
               kid: .keys[0].kid,
               e: .keys[0].e
               }
               
               " > rsa.jwk.pub
```

**Comment by a-mckinley** [comment]
> Or ... you install postgrest as a system service (https://docs.subzero.cloud/production-infrastructure/ubuntu-server/) and use systemd `ExecStartPre` directive to run one `curl` call that places that key in the file postgrest is configured to read...


Is there a straightforward way to make this happen with the docker image?

**Comment by savv** [comment]
Another solution for supporting firebase auth (or auth0), without worrying about JWKS support:
* put PostgREST behind nginx
* In nginx, use `auth_request` to have it reach out to a custom made backend that validates auth tokens (e.g. using firebase auth's admin lib); proxy responds with a `custom-proxy-user` response header
* Use `auth_request_set $auth_user $upstream_http_custom_proxy_user;` and `proxy_set_header` to pass it on to PostgREST
* use that instead of the JWT claim

Does this mean that JWT handling is another thing that PostgREST doesn't need to do? :-)

**Comment by pauldellinger** [comment]
How would you set a role without a JWT claim?

**Comment by savv** [comment]
I meant that tongue-in-cheek. You could simply pass the role as another, unsigned, HTTP header. Which, I think, would be fine in an environment where postgrest is behind nginx.

(In my case, I ended up generating a JWT in my auth proxy and passing that to postgrest, because JWTs are well integrated with the role system.)

**Comment by theonewolf** [comment]
It would be **really** awesome if we could load JWK key sets from URLs instead of local files such as:

https://www.googleapis.com/robot/v1/metadata/jwk/securetoken@system.gserviceaccount.com

This would prevent operator error in remembering to rotate those key sets once they expire as well---PostgREST should be able to reload them from the URL on expiration.

**Comment by steve-chavez** [comment]
For the meantime, @rockodragon shared a nodejs solution on gitter chat: https://gitter.im/begriffs/postgrest?at=5ea5106b568e5258e4852dbc

**Comment by theonewolf** [comment]
@rockodragon thanks!  I made something similar in Python, but don't have a great place to run it right now.

**Comment by saosebastiao** [comment]
Hi, just wanted to add a couple of major issues that I'm having with this particular limitation. 

Azure Active Directory (used internally where I work) uses a system of JWKSets which is available at a custom endpoint. For example, if I paste an AAD-issued jwt into jwt.io, what it does is it calls for the openid-configuration https://login.microsoftonline.com/blahblahblah/v2.0/.well-known/openid-configuration, which provides a `jwks_uri` attribute which can then be queried to get a JWKSet, which is just a collection of keys that can be used and cached. Then when it receives the token, it pulls a `kid` attribute from the JWT header section, and uses that valued to determine which key out of all of the collection of keys it needs to use. 

Personally I think it's pretty neat that you can get your keys this way, but the way that PostgREST gets its keys appears to be incompatible. This is because some of my users may end up having their JWT signed using a different key than others. Because PostgREST can only handle a single key at a time, as opposed to JWKSets, only some fraction of my users will be able to access my API, and the rest will fail with 401 errors. 

At least that is as far as I've gotten with this. If anybody knows some way to get Azure App Registration for AD to give out simple keys that can be used in PostgREST, I'm all ears. Apart from that, how difficult would an MR be to implement this for a someone with no Haskell experience? I've got a bit of Scala/OCaml experience, but when I looked at Auth.hs I think my brains fell out.

**Comment by saosebastiao** [comment]
Here is an example of another application working with this flow: 
https://www.voitanos.io/blog/validating-azure-ad-generated-oauth-tokens/

**Comment by bohanyang** [comment]
I'd like to share my solution. I wrote this to integrate PostgREST with ZITADEL (an open-source Auth0 alternative).
It's tricky that ZITADEL rotates the keys for every 6 hours by default, so the keys should be updated constantly.
I wrote a script to do that by updating the PostgREST setting in-database.

https://gist.github.com/bohanyang/aabb562ad12906ff13420a907c87d2ca

**Comment by hiteshjoshi** [comment]
I dont like the idea of using database for config.

So I did this

```docker
FROM debian:latest

# Install dependencies
RUN apt-get update && apt-get install -y \
    cron \
    supervisor \
    curl \
    jq \
    postgresql-client \
    postgrest

# Create necessary directories
RUN mkdir -p /var/log/supervisor /etc/supervisor/conf.d /app

# Copy the Supervisor configuration file
COPY supervisord.conf /etc/supervisor/conf.d/supervisord.conf

# Copy the cron job script
COPY fetch_jwk.sh /app/fetch_jwk.sh
RUN chmod +x /app/fetch_jwk.sh

# Copy the entrypoint script
COPY entrypoint.sh /app/entrypoint.sh
RUN chmod +x /app/entrypoint.sh

# Copy the crontab file
COPY crontab /etc/cron.d/fetch_jwk
RUN chmod 0644 /etc/cron.d/fetch_jwk

# Apply the cron job
RUN crontab /etc/cron.d/fetch_jwk

# Create the log file to be able to run tail
RUN touch /var/log/cron.log

# Entrypoint
ENTRYPOINT ["/app/entrypoint.sh"]
```


```
[supervisord]
nodaemon=true

[program:postgrest]
command=/usr/local/bin/postgrest /app/postgrest.conf
autostart=true
autorestart=true
stderr_logfile=/var/log/postgrest.err.log
stdout_logfile=/var/log/postgrest.out.log

[program:cron]
command=cron -f
autostart=true
autorestart=true
stderr_logfile=/var/log/cron.err.log
stdout_logfile=/var/log/cron.out.log
```

(fetch_jwk.sh)
```shell
#!/bin/sh

# Fetch the JWK secret from the Supertokens API server
JWK_RESPONSE=$(curl -s http://127.0.0.1:3000/auth/jwt/jwks.json)

# Escape the JWK response for the config file
ESCAPED_JWK_RESPONSE=$(echo "$JWK_RESPONSE" | jq -sRr @json)

# Update postgrest.conf
cat <<EOF > /app/postgrest.conf
db-uri = "$DATABASE_URL"
db-schema = "ploton_api"
db-anon-role = "web_anon"
jwt-secret = $ESCAPED_JWK_RESPONSE
server-port = 9000
server-host = "0.0.0.0"
server-cors-allowed-origins="$ALLOWED_CORS"
EOF

# Reload PostgREST
killall -SIGUSR2 postgrest
```


crontab
```
# Run fetch_jwk.sh every 6 days
0 0 */6 * * /app/fetch_jwk.sh >> /var/log/cron.log 2>&1
```


Basically updating the config file and then notifying postgrest about the config reload.

**Comment by hiteshjoshi** [comment]
oh nice. I didn't know about the filepath. Thanks for sharing.

**Comment by mike-eh2** [comment]
Thanks @hiteshjoshi. I can share an iteration on your solution that includes some simplifications:

Set `PGRST_JWT_SECRET` to `"@/app/jwk.pub"`.

Dockerfile
```
FROM postgrest/postgrest

USER root

RUN apt-get update && apt install -y curl cron psmisc

# Copy the cron job script
COPY update_jwk.sh /app/update_jwk.sh
RUN chmod +x /app/update_jwk.sh

# Copy the crontab file
COPY crontab /etc/cron.d/update_jwk
RUN chmod 0644 /etc/cron.d/update_jwk

# Apply the cron job
RUN crontab /etc/cron.d/update_jwk

# Get the current JWK
RUN ./app/update_jwk.sh
```
crontab
```
# Run update_jwk.sh at midnight
0 0 */1 * * /app/update_jwk.sh >> /var/log/cron.log 2>&1

```
update_jwk.sh
```
#!/bin/sh

# Fetch the public JWK
curl $JWK_URL > /app/jwk.pub

# Ask PostgREST to reload config
# Fails during Docker build because PostgREST isn't running yet
killall -SIGUSR2 postgrest || true

```

**Comment by atatdotdot** [comment]
@steve-chavez Would it align with the ethos of the project to include dynamic JWKS loading from a well-known endpoint directly in PostgREST? Or is it preferable for such functionality to handled as an external exercise as implemented here?

**Comment by monneyboi** [comment]
> To be honest, I don't see the value, really.

In my case, the added value would be being able to just set some environment vars & run PostgREST. Now i have to add [some cruft](https://github.com/resolve-works/insight/blob/master/docker-compose.yml#L70) in my `docker-compose.yml` file to get it working. 

I see less value in automatically refreshing, but the loading part would save everybody that uses PostgREST with JWKS a trip to this here ticket.

**Comment by rudetrue** [comment]
Automatically refreshing would allow rotating keys without restarting too. It seems like most things have standardized on fetching the JWK for this reason, from what I've seen.

**Comment by rudetrue** [comment]
Yes but it's not easy. I spent 2-4 hours today trying to figure out the right `jq` incantation to parse the JWKS endpoint in my `curl` command. On top of adding a container just to do this task on PostgREST startup.

**Comment by monneyboi** [comment]
I believe this is handled through a [openid configuration request](
https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig)

**Comment by monneyboi** [comment]
Yeah, querying the OpenID connect endpoint seems to be the way others are handling this. For example: [RabbitMQ](https://www.rabbitmq.com/docs/oauth2#authorization-flow), [OpenSearch](https://opensearch.org/docs/latest/security/authentication-backends/openid-connect/#configure-openid-connect-integration), [Minio](https://min.io/docs/minio/linux/operations/external-iam/configure-openid-external-identity-management.html#set-the-openid-configuration-settings).

**Comment by hmoazzem** [comment]
I've been trying to figure a way out, and came up with https://github.com/edgeflare/edge/blob/main/internal/util/postgrest/rotate-jwt.go.go and calling it like

```go
func main() {
	ctx := context.Background()
	pool, _ := pgxpool.New(ctx, os.Getenv("CONN_STRING"))
	err := postgrest.RotateJwtKey(ctx, pool, os.Getenv("ZITADEL_JWK_URL"))
	if err != nil {
		log.Fatalln(err)
	}
}
```

**Comment by BergCyrill** [comment]
Is there any progress on this? Currently I am trying to figure out how to configure it properly when using Dex IdP and haven't found a way where you would not have severe service interruptions. Per default dex rotates the signing key every 6 hours but even when setting it to a longer timeframe, postgrest would reject jwt which were signed between two cron job runs after the rotation has happened. Or am I missing something here?

**Comment by BergCyrill** [comment]
I maybe described my issue (which I think is the exact same issue others have on this GitHub issue) a bit too  complex, sorry for that. I am totally aware of jwks and the possibility to configure them in postgrest - that's what I am actually doing. The best solution so far is to have some kind of job/tool running every 5 seconds to fetch the jwks from dex, compare it to the one configured in postgrest and change it if different + reload postgrest configuration (like this one https://gist.github.com/bohanyang/aabb562ad12906ff13420a907c87d2ca).
Having to build this job/tool and additionally build extra clients which will just retry after 5-10 seconds in case of a "JWS Invalid signature" because it is possible that dex has just rotated the signing key but the job has not updated the config so far feels somewhat hacky.
And I think it is a lot of overhead (and prone for errors) for something that could be solved better if postgrest would be able to retrieve its jwks from an url (maybe every time it is confronted with a jwt where signing cannot be verified and some rate limiting to mitigate malicious requests).

**Comment by taimoorzaeem** [comment]
Now that we have a built-in HTTP client, maybe we can work on a design for this feature?

**Comment by sweco-sedalh** [comment]
> I have no idea how existing systems handle this though.

I believe a common pattern is to check the `kid` claim in the JWT header - if the client (PostgREST in this case) doesn't have a key stored with that `kid` it reloads keys from the IdP (the list of `kid`s that have been attempted to fetch should also be stored to prevent a DoS attack).

See e.g. https://github.com/jpadilla/pyjwt/blob/5b38225f16d9d03a51056221a1d3d30b84b69fef/jwt/jwks_client.py#L170-L211

**Comment by savv** [comment]
Is this fixed with #5266 then? Which version could we expect it to land in?
